Split the rename function according to the level and pass expression by copy instead of reference

romainbrenguier opened the pull request:
~This is based on https://github.com/diffblue/cbmc/pull/3988 and~ This is an intermediary step towards https://github.com/diffblue/cbmc/pull/3986
We split the rename functions according to there level, so that we can later make them return types which reflects the transformation.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Make level argument of rename a template argument

The level argument is always known from the call site (except for
rename recursive calls), and there are only 3 possible values, so it
makes sense to have it as a template argument.
This will be useful in particular, when we want the return type to
depend on the level.

romainbrenguier committed:
Extract a rename_level0_ssa function

The rename<L0> function will normally be used with ssa symbols, we add
that function to avoid unecessary decision making.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -280,14 +280,12 @@ static void set_l2_indices(
 }
 
-void goto_symex_statet::rename_level0(exprt &expr, const namespacet &ns)
+exprt goto_symex_statet::rename_level0(exprt expr, const namespacet &ns)
```
The parameter should be `const exprt &expr`

romainbrenguier replied:
In the follow up PR I will make this return another type, and then having the argument passed by copy allow to move the argument, potentially saving a copy. I also refactored the code so that this will not get called recursively.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -280,6 +280,46 @@ static void set_l2_indices(
   ssa_expr.set_level_2(level2.current_count(ssa_expr.get_identifier()));
 }
 
+void goto_symex_statet::rename_level0(exprt &expr, const namespacet &ns)
```
I would expect that adding this function goes together with removing the same code from another function. As-is, it's difficult to assess whether this code does the same as the old one.

romainbrenguier committed:
Extract a rename_level1_ssa function

This can be useful when we now from the caller that the expression is
already is in SSA form.

romainbrenguier committed:
Use rename_level1_ssa where possible

In places where rename<L1> is known to be called on an ssa_exprt, we
replace the call by a call to rename_level1_ssa to avoid unecessary
decision making and show more information in the return type.

romainbrenguier committed:
Make rename take a copy instead of reference

This makes the interface clearer as there is no argument that is both an
input and an output.
This also corresponds to how the function was used in a lot of
occurences, in which case the code is now simpler.
In the cases where we actually want the transformation to happen
in-place, we can use std::move.

romainbrenguier commented:
@tautschnig @smowton this is ready for review

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -261,97 +262,160 @@ void goto_symex_statet::set_l2_indices(
+{
+  renaming_strategyt strategy;
+  strategy.rename_ssa = [this](ssa_exprt ssa, const namespacet &ns) -> exprt {
+    set_l0_indices(ssa, ns);
```
Surely `rename_level0_ssa`

romainbrenguier replied:
done

romainbrenguier replied:
done

smowton reviewed (commented):
The strategy pattern seems very verbose compared to just passing `levelt`, since 2/3 of its members are always the same and the other one only ever differs by level.

romainbrenguier commented:
> The strategy pattern seems very verbose compared to just passing `levelt`, 

True this is more verbose, the main intention was to avoid late decision making in the rename function, we already know what the level is, so there is no reason to add more ifs in the loop. 
I also find it more structured that way but that's a matter of taste. 

> since 2/3 of its members are always the same and the other one only ever differs by level.

Note that the members of the strategy are not exactly the same since the level argument differs, and we could think of also splitting the called functions according to their level.

tautschnig reviewed (changes requested):
I have only started reviewing this and I'm afraid the first commit is a deal breaker: "Refactor goto_symex_state::rename to use a strategy" - there are some very good changes in there (such as nicely introducing the depth iterator), but these are completely unrelated to introducing a strategy. If such changes are a necessary precondition to introducing a strategy then they must go in a separate commit, quite possibly even a separate PR. This is an area of code that is crucial both in terms of performance and correctness. We must not have multiple changes in a single commit as it prohibits identifying both performance and correctness issues or advances. Yes, we also need to be able to pinpoint improvements. Let's make such changes in isolation, benchmark them, and then follow up with the next change.

romainbrenguier commented:
@tautschnig 
> I have only started reviewing this and I'm afraid the first commit is a deal breaker: "Refactor goto_symex_state::rename to use a strategy" - there are some very good changes in there (such as nicely introducing the depth iterator), but these are completely unrelated to introducing a strategy. If such changes are a necessary precondition to introducing a strategy then they must go in a separate commit, quite possibly even a separate PR. This is an area of code that is crucial both in terms of performance and correctness. We must not have multiple changes in a single commit as it prohibits identifying both performance and correctness issues or advances. Yes, we also need to be able to pinpoint improvements. Let's make such changes in isolation, benchmark them, and then follow up with the next change.

That's a good idea indeed. I tried to split the first commit, and started to profile the changes. It turns out changing the recursive function into one using iterators is not such a good idea. Just making it non-recursive (without the use of strategies) makes it significantly slower. I thought I would be able to save expression writes that way, but I think `mutate` is an issue here as it will detach all the expressions in the stack for each mutate. 
I think I will redo these changes, keeping the function recursive, and using a template argument instead of strategies.
If we want to minimize the number of expression write, I think we need to make `rename(typet ...)` return an optional so that we know wether it needs to be updated.

tautschnig commented:
> I tried to split the first commit, and started to profile the changes. It turns out changing the recursive function into one using iterators is not such a good idea.

Unless you have binned it already and `git reflog` can't recover it either, would you still be ok to create a pull request just with this change? I'd at least try to discuss it. Notably, the cost may also just be distributed differently.

romainbrenguier commented:
@tautschnig 
> Unless you have binned it already and git reflog can't recover it either, would you still be ok to create a pull request just with this change? I'd at least try to discuss it. Notably, the cost may also just be distributed differently.

Pull request here: https://github.com/diffblue/cbmc/pull/4254

romainbrenguier commented:
@tautschnig I rewrote the PR to use templates instead of strategies. This is ready for review.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 659d220c20b55d5e17136af346b5aed6d2b63cc0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102171205
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

smowton reviewed (approved):
This looks great -- one suggestion, since a fair majority of renamings are to L2, maybe use a default template parameter much as there is currently a default ordinary parameter?

tautschnig commented on src/goto-symex/symex_decl.cpp:
```diff
@@ -69,7 +69,11 @@ void goto_symext::symex_decl(statet &state, const symbol_exprt &expr)
+  exprt expr_l2 = state.rename<statet::L2>(std::move(ssa), ns);
+  INVARIANT(
+    expr_l2.id() == ID_symbol && expr_l2.get_bool(ID_C_SSA_symbol),
+    "symbol to declare should not be replaced by constant propagation");
```
Fair enough, but unrelated to the other changes here?

romainbrenguier replied:
We need these conditions to be sure `to_ssa_expr` will not failed in the next line. Before `ssa` was modified in-place so there was no conversion here.

tautschnig replied:
Ah, indeed that possible problem was hidden before. But `to_ssa_expr` should already do the same checking for you?

tautschnig commented on src/goto-symex/symex_builtin_functions.cpp:
```diff
@@ -481,11 +475,8 @@ void goto_symext::symex_trace(
-      vars.push_back(var);
-    }
+      vars.push_back(
+        state.rename<goto_symex_statet::L2>(code.arguments()[j], ns));
```
Nit pick: keep the braces.

tautschnig commented on src/goto-symex/symex_builtin_functions.cpp:
```diff
@@ -359,18 +355,16 @@ void goto_symext::symex_output(
-    do_simplify(args.back());
+    exprt l2_arg = state.rename<goto_symex_statet::L2>(args.back(), ns);
+    do_simplify(l2_arg);
+    args.back() = std::move(l2_arg);
```
I think you should get rid of `args.push_back(code.operands()[i])`

tautschnig commented on src/goto-symex/symex_builtin_functions.cpp:
```diff
@@ -336,17 +333,16 @@ void goto_symext::symex_input(
-    do_simplify(args.back());
+    exprt l2_arg = state.rename<goto_symex_statet::L2>(args.back(), ns);
+    do_simplify(l2_arg);
+    args.back() = std::move(l2_arg);
```
Get rid of `args.push_back(...)`

tautschnig commented on src/goto-symex/symex_builtin_functions.cpp:
```diff
@@ -311,11 +309,10 @@ void goto_symext::symex_printf(
   do_simplify(tmp_rhs);
 
-  const exprt::operandst &operands=tmp_rhs.operands();
+  const exprt::operandst &operands = tmp_rhs.operands();
```
Unnecessary change.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -727,7 +724,10 @@ void goto_symex_statet::rename(
-        rename<level>(to_array_type(component.type()).size(), ns);
+      {
+        auto &array_type = to_array_type(component.type());
+        array_type.size() = rename<level>(array_type.size(), ns);
```
Above you use `std::move(array_type.size())` - please be consistent.

tautschnig reviewed (changes requested):
I think this mostly looks good, but I strongly dislike that users now have two seemingly very different APIs to do L0/L1 renaming. Couldn't we have `template<T> ssa_exprt rename(ssa_exprt, namespacet)` so that the compiler would just choose the most appropriate variant for you?

romainbrenguier commented:
> I think this mostly looks good, but I strongly dislike that users now have two seemingly very different APIs to do L0/L1 renaming. Couldn't we have `template<T> ssa_exprt rename(ssa_exprt, namespacet)` so that the compiler would just choose the most appropriate variant for you?

I think this will make it ambiguous because `ssa_exprt` is also an `exprt`, but I guess nothing wrong can happen. I will try your solution.

romainbrenguier commented:
@tautschnig 
> I think this mostly looks good, but I strongly dislike that users now have two seemingly very different APIs to do L0/L1 renaming. Couldn't we have template<T> ssa_exprt rename(ssa_exprt, namespacet) so that the compiler would just choose the most appropriate variant for you?

> I think this will make it ambiguous because ssa_exprt is also an exprt, but I guess nothing wrong can happen. I will try your solution.

The problem is that there is no equivalent of this function for L2 (it would not always return an ssa_exprt  because of constant propagation). So either we will get an template that only works for level=L0, or L1 which I think is confusing. Or we need to return a type which depends on the level, which will involve a bit more template coding. What do you think?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3333db08bbfcfdb1603fb87dd8f328c4fe3ed457).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102350102

tautschnig commented:
> The problem is that there is no equivalent of this function for L2 (it would not always return an ssa_exprt because of constant propagation). So either we will get an template that only works for level=L0, or L1 which I think is confusing. Or we need to return a type which depends on the level, which will involve a bit more template coding. What do you think?

Isn't it possible to delete a specialization? I think you should be able to do
````c++
template<> ssa_exprt rename<L2>(ssa_exprt, const namespacet &) = delete;
````
and the compiler should then happily use `exprt rename<L2>(...)` even when you pass in an `ssa_exprt` as first argument. But maybe I'm too optimistic about C++ here?

romainbrenguier commented:
@tautschnig 
> Isn't it possible to delete a specialization? I think you should be able to do
> 
 ```c++
template<> ssa_exprt rename<L2>(ssa_exprt, const namespacet &) = delete;
```
> and the compiler should then happily use `exprt rename<L2>(...)` even when you pass in an `ssa_exprt` as first argument. But maybe I'm too optimistic about C++ here? use of deleted function 

In that case I get the error: `use of deleted function ‘ssa_exprt goto_symex_statet::rename(ssa_exprt, const namespacet&) [with goto_symex_statet::levelt level = (goto_symex_statet::levelt)2u]’`
when calling `rename<L2>`

tautschnig commented:
> In that case I get the error: `use of deleted function ‘ssa_exprt goto_symex_statet::rename(ssa_exprt, const namespacet&) [with goto_symex_statet::levelt level = (goto_symex_statet::levelt)2u]’`
when calling `rename<L2>`

Hmm, too bad. Maybe we should just not use templates here at all? Just have functions `rename_l0`, `rename_l1`, `rename_l2`? For `rename_l0` and `rename_l1` we would then have polymorphic variants  `exprt -> exprt` and `ssa_exprt -> ssa_exprt`, while for `rename_l2` we would only have `exprt -> exprt`, or maybe `ssa_exprt -> exprt`?

romainbrenguier commented:
@tautschnig 
> Hmm, too bad. Maybe we should just not use templates here at all? Just have functions rename_l0, rename_l1, rename_l2? For rename_l0 and rename_l1 we would then have polymorphic variants exprt -> exprt and ssa_exprt -> ssa_exprt, while for rename_l2 we would only have exprt -> exprt, or maybe ssa_exprt -> exprt?

The goal in having a template for the non SSA case was to avoid code duplication, because the version for three levels have a lot in common.
I think the current version of the interface is fine, the user can always use `rename<level>`, and the `rename_level0/1_ssa` is there only if they want to avoid using `to_ssa_expr` on the return value (and avoid some unnecessary condition checks in the called function).
I could replace the two functions by a template `rename_ssa<levelt>` if you think that's better (the L2 version would be deleted).

tautschnig commented:
> I could replace the two functions by a template `rename_ssa<levelt>` if you think that's better (the L2 version would be deleted).

Ok, that's probably the best approach. It makes functions that do the same (though slightly more performant for `ssa_exprt`) (almost) look the same.

romainbrenguier committed:
Merge the two rename_ssa functions

We merge rename_level0_ssa and rename_level1_ssa in one template
rename_ssa<level> since the code is similar and their usage would be
similar.

romainbrenguier commented:
@tautschnig @smowton I added a commit to merge the SSA function into one template

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a3fe46733bb36fb5286020fd328af1bf41c9d392).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102488874

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -330,13 +330,9 @@ void goto_symex_statet::rename(exprt &expr, const namespacet &ns)
-    expr=ssa_exprt(expr);
-    rename<level>(expr, ns);
+    else
+      expr = rename<level>(ssa_exprt(expr), ns);
```
Nit pick: you otherwise aim to use `{...}` in constructor calls, so do `ssa_exprt{expr}` here as well.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -263,24 +263,29 @@ void goto_symex_statet::set_l2_indices(
+  else if(level == L1)
+    set_l1_indices(ssa, ns);
+  else
+    UNREACHABLE;
```
How about doing a full specialization here? It would do away with the `static_assert` and all the branching (that really isn't one, because the compiler has to remove it anyway). But that's a nit-pick, up to you.

romainbrenguier replied:
This version avoids a bit a code duplication, so I would leave it like that. I think the static assert gives an error message that is clearer than getting a linker error.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -331,8 +336,10 @@ exprt goto_symex_statet::rename(exprt expr, const namespacet &ns)
+    else if(level == L0 || level == L1)
       expr = rename<level>(ssa_exprt(expr), ns);
+    else // No SSA version for level2
+      expr = rename<level>((exprt)ssa_exprt(expr), ns);
```
Is this still necessary, given they ssa vs. non-ssa functions have different names? It if is, use `static_cast`, not `(exprt)`.

tautschnig commented on src/goto-symex/symex_function_call.cpp:
```diff
@@ -395,8 +395,8 @@ static void locality(
-    ssa_exprt ssa =
-      state.rename_level0_ssa(ssa_exprt(ns.lookup(*it).symbol_expr()), ns);
+    ssa_exprt ssa = state.rename_ssa<goto_symex_statet::L0>(
+      ssa_exprt(ns.lookup(*it).symbol_expr()), ns);
```
Use `ssa_exprt{...}`

tautschnig commented:
Any idea about performance differences (if any)?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7da918f2b8fd8aefe2acc583828d718da9af35ae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102504151

romainbrenguier commented:
@tautschnig 
> Any idea about performance differences (if any)?

I ran symex on 70 functions, the time before the changes was `102.72` and after `100.88`.
